src/libgit2/fetch.c

```diff
@@ -173,31 +176,55 @@ int git_fetch_negotiate(git_remote *remote, const git_fetch_options *opts)
 
 	remote->need_pack = 0;
 
+	if (!opts)
+		remote->nego.depth = -1;
+	else
+		remote->nego.depth = opts->unshallow ? INT_MAX : opts->depth;
+
 	if (filter_wants(remote, opts) < 0)
 		return -1;
 
 	/* Don't try to negotiate when we don't want anything */
 	if (!remote->need_pack)
 		return 0;
 
+	if (opts && opts->unshallow && opts->depth > 0) {
+		git_error_set(GIT_ERROR_INVALID, "options '--depth' and '--unshallow' cannot be used together");
+		return -1;
+	}
+
 	/*
 	 * Now we have everything set up so we can start tell the
 	 * server what we want and what we have.
 	 */
+	remote->nego.refs = (const git_remote_head * const *)remote->refs.contents;
+	remote->nego.count = remote->refs.length;
+	remote->nego.shallow_roots = git__malloc(sizeof(git_shallowarray));
```

Comment:
```suggestion
	remote->nego.shallow_roots = git__malloc(sizeof(*remote->nego.shallow_roots));
```

Again, not sure if there are project guidelines for `malloc()` but this is more future proof.